staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1607 @@
+            format: req.format,
+            bind: RUTABAGA_PIPE_BIND_RENDER_TARGET,
+            width: req.width,
+            height: req.height,
+            depth: 1,
+            array_size: 1,
+            last_level: 0,
+            nr_samples: 0,
+            flags: 0,
+        };
+        virtio_gpu.resource_create_3d(req.resource_id, resource_create_3d)
+    }
+
+    fn handle_transfer_to_host_2d(
+        virtio_gpu: &mut impl VirtioGpu,
+        req: virtio_gpu_transfer_to_host_2d,
+    ) -> VirtioGpuResult {
+        let transfer = Transfer3D::new_2d(req.r.x, req.r.y, req.r.width, req.r.height, req.offset);
+        virtio_gpu.transfer_write(0, req.resource_id, transfer)
+    }
+
+    fn handle_update_cursor(
+        virtio_gpu: &mut impl VirtioGpu,
+        req: virtio_gpu_update_cursor,
+    ) -> VirtioGpuResult {
+        let cursor_pos = VhostUserGpuCursorPos {
+            scanout_id: req.pos.scanout_id,
+            x: req.pos.x,
+            y: req.pos.y,
+        };
+        virtio_gpu.update_cursor(req.resource_id, cursor_pos, req.hot_x, req.hot_y)
+    }
+
+    fn handle_move_cursor(
+        virtio_gpu: &mut impl VirtioGpu,
+        req: virtio_gpu_update_cursor,
+    ) -> VirtioGpuResult {
+        let cursor = VhostUserGpuCursorPos {
+            scanout_id: req.pos.scanout_id,
+            x: req.pos.x,
+            y: req.pos.y,
+        };
+        virtio_gpu.move_cursor(req.resource_id, cursor)
+    }
+
+    fn handle_ctx_create(
+        virtio_gpu: &mut impl VirtioGpu,
+        hdr: virtio_gpu_ctrl_hdr,
+        req: virtio_gpu_ctx_create,
+    ) -> VirtioGpuResult {
+        let context_name: Option<String> = String::from_utf8(req.debug_name.to_vec()).ok();
+        virtio_gpu.create_context(hdr.ctx_id, req.context_init, context_name.as_deref())
+    }
+
+    fn handle_resource_create_3d(
+        virtio_gpu: &mut impl VirtioGpu,
+        req: virtio_gpu_resource_create_3d,
+    ) -> VirtioGpuResult {
+        let resource_create_3d = ResourceCreate3D {
+            target: req.target,
+            format: req.format,
+            bind: req.bind,
+            width: req.width,
+            height: req.height,
+            depth: req.depth,
+            array_size: req.array_size,
+            last_level: req.last_level,
+            nr_samples: req.nr_samples,
+            flags: req.flags,
+        };
+        virtio_gpu.resource_create_3d(req.resource_id, resource_create_3d)
+    }
+
+    fn handle_transfer_to_host_3d(
+        virtio_gpu: &mut impl VirtioGpu,
+        ctx_id: u32,
+        req: virtio_gpu_transfer_host_3d,
+    ) -> VirtioGpuResult {
+        let transfer = Transfer3D {
+            x: req.box_.x,
+            y: req.box_.y,
+            z: req.box_.z,
+            w: req.box_.w,
+            h: req.box_.h,
+            d: req.box_.d,
+            level: req.level,
+            stride: req.stride,
+            layer_stride: req.layer_stride,
+            offset: req.offset,
+        };
```

Comment:
Idea for future work not necessarily this PR: add `From/Into` conversions for these types